File: source/python/fan_out_fan_in.py
import threading
import queue

work_q = queue.Queue()
result_q = queue.Queue()

def worker():
    while True:
        val = work_q.get()
        if val is None:
            break
        result_q.put(val * val)

for i in range(1, 21):
    work_q.put(i)

workers = []
for _ in range(4):
    t = threading.Thread(target=worker)
    t.start()
    workers.append(t)

for _ in range(4):
    work_q.put(None)

for t in workers:
    t.join()

results = []
while not result_q.empty():
    results.append(result_q.get())

results.sort()
print(f"results: {' '.join(str(x) for x in results)}")

File: source/rust/fan_out_fan_in.rs
use std::sync::mpsc;
use std::thread;

fn main() {
    let (work_tx, work_rx) = mpsc::channel::<i32>();
    let work_rx = std::sync::Arc::new(std::sync::Mutex::new(work_rx));
    let (result_tx, result_rx) = mpsc::channel::<i32>();

    for i in 1..=20 {
        work_tx.send(i).unwrap();
    }
    drop(work_tx);

    let mut workers = Vec::new();
    for _ in 0..4 {
        let rx = work_rx.clone();
        let tx = result_tx.clone();
        workers.push(thread::spawn(move || {
            loop {
                let val = { rx.lock().unwrap().recv() };
                match val {
                    Ok(v) => tx.send(v * v).unwrap(),
                    Err(_) => break,
                }
            }
        }));
    }
    drop(result_tx);

    for w in workers { w.join().unwrap(); }

    let mut results: Vec<i32> = result_rx.into_iter().collect();
    results.sort();
    let s: Vec<String> = results.iter().map(|x| x.to_string()).collect();
    println!("results: {}", s.join(" "));
}
